libraries/AC_Autorotation/AC_Autorotation.cpp

```diff
@@ -2,375 +2,524 @@
+    // Get singleton for RPM library
+    const AP_RPM *rpm = AP_RPM::get_singleton();
 
-    //Limits the maximum change in pitch attitude based on acceleration
-    if (_accel_target > _accel_out_last + _accel_max) {
-        _accel_target = _accel_out_last + _accel_max;
-    } else if (_accel_target < _accel_out_last - _accel_max) {
-        _accel_target = _accel_out_last - _accel_max;
+    // Get current rpm, checking to ensure no nullptr
+    if (rpm == nullptr) {
+        hal.util->snprintf(buffer, buflen, "Can't access RPM");
+        return false;
     }
 
-    //Limiting acceleration based on velocity gained during the previous time step 
-    if (fabsf(_delta_speed_fwd) > _accel_max * _dt) {
-        _flag_limit_accel = true;
-    } else {
-        _flag_limit_accel = false;
+    // Sanity check that the designated rpm sensor instance is there
+    if ((_param_rpm_instance.get() < 0)) {
+        hal.util->snprintf(buffer, buflen, "RPM instance <0");
+        return false;
     }
 
-    if ((_flag_limit_accel && fabsf(_accel_target) < fabsf(_accel_out_last)) || !_flag_limit_accel) {
-        _accel_out = _accel_target;
-    } else {
-        _accel_out = _accel_out_last;
+    if (!rpm->enabled(_param_rpm_instance.get())) {
+        hal.util->snprintf(buffer, buflen, "RPM%i not enabled", _param_rpm_instance.get()+1);
+        return false;
     }
-    _accel_out_last = _accel_out;
+#endif
 
-    // update angle targets that will be passed to stabilize controller
-    _pitch_target = accel_to_angle(-_accel_out*0.01) * 100;
+    // Check that heli motors is configured for autorotation
+    if (!_motors_heli->rsc_autorotation_enabled()) {
+        hal.util->snprintf(buffer, buflen, "H_RSC_AROT_* not configured");
+        return false;
+    }
 
+    return true;
 }
 
-
-// Determine the forward ground speed component from measured components
-float AC_Autorotation::calc_speed_forward(void)
+// Check if we believe we have landed. We need the landed state to zero all
+// controls and make sure that the copter landing detector will trip
+bool AC_Autorotation::check_landed(void)
 {
-    auto &ahrs = AP::ahrs();
-    Vector2f groundspeed_vector = ahrs.groundspeed_vector();
-    float speed_forward = (groundspeed_vector.x*ahrs.cos_yaw() + groundspeed_vector.y*ahrs.sin_yaw())* 100; //(c/s)
-    return speed_forward;
+    // minimum speed (m/s) used for "is moving" check
+    const float min_moving_speed = 1.0;
+
+    Vector3f velocity;
+    const AP_AHRS &ahrs = AP::ahrs();
+    _landed_reason.min_speed = ahrs.get_velocity_NED(velocity) && (velocity.length() < min_moving_speed);
+    _landed_reason.land_col = _motors_heli->get_below_land_min_coll();
+    _landed_reason.is_still = AP::ins().is_still();
+
+    return _landed_reason.min_speed && _landed_reason.land_col && _landed_reason.is_still;
 }
 
+// Dynamically update time step used in autorotation controllers
+void AC_Autorotation::set_dt(float delta_sec)
+{
+    if (is_positive(delta_sec)) {
```

Comment:
We don't bother with this check in the attitude controller, I think its safe to assume DT is valid.